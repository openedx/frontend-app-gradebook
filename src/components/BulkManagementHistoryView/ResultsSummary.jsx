/* eslint-disable react/sort-comp, react/button-has-type */
import React from 'react';
import PropTypes from 'prop-types';

import { Hyperlink, Icon } from '@edx/paragon';
import { Download } from '@edx/paragon/icons';

import lms from 'data/services/lms';

/**
 * <ResultsSummary {...{ courseId, rowId, text }} />
 * displays a result summary cell for a single bulk management upgrade history entry.
 * @param {string} courseId - course identifier
 * @param {number} rowId - row/error identifier
 * @param {string} text - summary string
 */
const ResultsSummary = ({
  rowId,
  text,
}) => (
  <Hyperlink
    href={lms.urls.bulkGradesUrlByRow(rowId)}
    destination="www.edx.org"
    target="_blank"
    rel="noopener noreferrer"
    showLaunchIcon={false}
  >
    <Icon src={Download} className="d-inline-block" />
    {text}
  </Hyperlink>
);

ResultsSummary.propTypes = {
  rowId: PropTypes.number.isRequired,
  text: PropTypes.string.isRequired,
};

export default ResultsSummary;
