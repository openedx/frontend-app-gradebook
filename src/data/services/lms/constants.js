import { StrictDict } from '../../../utils';

export const pageSize = 25;
export const historyRecordLimit = 5;

export const paramKeys = StrictDict({
  cohortId: 'cohort_id',
  pageSize: 'page_size',
  userContains: 'user_contains',
  enrollmentMode: 'enrollment_mode',
  assignment: 'assignment',
  assignmentGradeMin: 'assignment_grade_min',
  assignmentGradeMax: 'assignment_grade_max',
  courseGradeMin: 'course_grade_min',
  courseGradeMax: 'course_grade_max',
  excludedCourseRoles: 'excluded_course_roles',
});
