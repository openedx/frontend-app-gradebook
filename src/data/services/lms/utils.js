import queryString from 'query-string';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { filters } from '@src/data/constants/filters';

/**
 * get(url)
 * simple wrapper providing an authenticated Http client get action
 * @param {string} url - target url
 */
export const get = (...args) => getAuthenticatedHttpClient().get(...args);
/**
 * post(url, data)
 * simple wrapper providing an authenticated Http client post action
 * @param {string} url - target url
 * @param {object|string} data - post payload
 */
export const post = (...args) => getAuthenticatedHttpClient().post(...args);

/**
 * stringifyUrl(url, query)
 * simple wrapper around queryString.stringifyUrl that sets skip behavior
 * @param {string} url - base url string
 * @param {object} query - query parameters
 */
export const stringifyUrl = (url, query) => queryString.stringifyUrl(
  { url, query },
  { skipNull: true, skipEmptyString: true },
);

/**
 * filterQuery(options)
 * Takes current filter object and returns it with only valid filters that are
 * set and have non-'All' values
 * @param {object} options - filter values
 * @return {object} - valid filters that are set and do not equal 'All'
 */
export const filterQuery = (options) => Object.values(filters)
  .filter(filter => options[filter] && options[filter] !== 'All')
  .reduce(
    (obj, filter) => ({ ...obj, [filter]: options[filter] }),
    {},
  );
