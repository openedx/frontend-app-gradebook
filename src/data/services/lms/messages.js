import { StrictDict } from '../../../utils';

export default StrictDict({
  errors: {
    missingAssignment: (
      'Gradebook LMS API requires assignment to be set to filter by min/max assig. grade'
    ),
    unhandledResponse: 'unhandled response error',
  },
});
