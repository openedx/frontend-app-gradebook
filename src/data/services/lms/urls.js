import { getConfig } from '@edx/frontend-platform';
import { StrictDict } from 'utils';
import { historyRecordLimit } from './constants';
import { filterQuery, stringifyUrl } from './utils';

const courseId = window.location.pathname.split('/').filter(Boolean).pop() || '';

export const getUrlPrefix = () => `${getConfig().LMS_BASE_URL}/api/`;
export const getBulkGradesUrl = () => `${getUrlPrefix()}bulk_grades/course/${courseId}/`;
export const getEnrollmentUrl = () => `${getUrlPrefix()}enrollment/v1/`;
export const getGradesUrl = () => `${getUrlPrefix()}grades/v1/`;
export const getGradebookUrl = () => `${getGradesUrl()}gradebook/${courseId}/`;
export const getBulkUpdateUrl = () => `${getGradebookUrl()}bulk-update`;
export const getInterventionUrl = () => `${getBulkGradesUrl()}intervention/`;
export const getCohortsUrl = () => `${getUrlPrefix()}courses/${courseId}/cohorts/`;
export const getTracksUrl = () => `${getEnrollmentUrl()}course/${courseId}?include_expired=1`;
export const getBulkHistoryUrl = () => `${getBulkUpdateUrl()}history/`;
export const getAssignmentTypesUrl = () => stringifyUrl(`${getGradebookUrl()}grading-info`, { graded_only: true });
export const getRolesUrl = () => stringifyUrl(`${getEnrollmentUrl()}roles/`, { courseId });
/**
 * bulkGradesUrlByCourseAndRow(courseId, rowId)
 * returns the bulkGrades url with the given rowId.
 * @param {string} rowId - row/error identifier
 * @return {string} - bulk grades fetch url
 */
export const bulkGradesUrlByRow = (rowId) => stringifyUrl(getBulkGradesUrl(), { error_id: rowId });

export const gradeCsvUrl = (options = {}) => stringifyUrl(getBulkGradesUrl(), filterQuery(options));

export const interventionExportCsvUrl = (options = {}) => (
  stringifyUrl(getInterventionUrl(), filterQuery(options))
);

export const sectionOverrideHistoryUrl = (subsectionId, userId) => stringifyUrl(
  `${getGradesUrl()}subsection/${subsectionId}/`,
  { user_id: userId, history_record_limit: historyRecordLimit },
);

export default StrictDict({
  getUrlPrefix,
  getBulkGradesUrl,
  getEnrollmentUrl,
  getGradesUrl,
  getGradebookUrl,
  getBulkUpdateUrl,
  getInterventionUrl,
  getCohortsUrl,
  getTracksUrl,
  getBulkHistoryUrl,
  getAssignmentTypesUrl,
  getRolesUrl,
  bulkGradesUrlByRow,
  gradeCsvUrl,
  interventionExportCsvUrl,
  sectionOverrideHistoryUrl,
});
