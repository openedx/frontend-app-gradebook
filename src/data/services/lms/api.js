import { StrictDict } from 'utils';

import urls, {
  gradeCsvUrl,
  sectionOverrideHistoryUrl,
} from './urls';
import { pageSize, paramKeys } from './constants';
import messages from './messages';

import * as utils from './utils';

const { get, post, stringifyUrl } = utils;

/*********************************************************************************
 * GET Actions
 *********************************************************************************/
const assignmentTypes = () => get(urls.getAssignmentTypesUrl());
const cohorts = () => get(urls.getCohortsUrl());
const roles = () => get(urls.getRolesUrl());
const tracks = () => get(urls.getTracksUrl());

/**
 * fetch.gradebookData(searchText, cohort, track, options)
 * fetches updated gradebook data based on current filter selections.
 * Raises an error if assignment grade limits are set, but not assignment.
 * @param {string} searchText - search text filter
 * @param {nunber} cohort - selected cohort filter
 * @param {string} track - selected track filter
 * @param {object} options - additional optional filter values
 * @return {Promise} - get response
 */
const gradebookData = (searchText, cohort, track, options = {}) => {
  if ((options.assignmentGradeMax || options.assignmentGradeMin) && !options.assignment) {
    throw new Error(messages.errors.missingAssignment);
  }
  const queryParams = {
    [paramKeys.pageSize]: pageSize,
    [paramKeys.userContains]: searchText,
    [paramKeys.cohortId]: cohort,
    [paramKeys.enrollmentMode]: track,
    [paramKeys.courseGradeMax]: options.courseGradeMax,
    [paramKeys.courseGradeMin]: options.courseGradeMin,
    [paramKeys.excludedCourseRoles]: options.includeCourseRoleMembers ? null : ['all'],
    [paramKeys.assignment]: options.assignment,
    [paramKeys.assignmentGradeMax]: options.assignmentGradeMax,
    [paramKeys.assignmentGradeMin]: options.assignmentGradeMin,
  };
  return get(stringifyUrl(urls.getGradebookUrl(), queryParams));
};

/**
 * fetch.gradeBulkOperationHistory()
 * fetches bulk operation history and raises an error if the operation fails
 * @return {Promise} - get response
 */
const gradeBulkOperationHistory = () => get(urls.getBulkHistoryUrl())
  .then(response => response.data)
  .catch(() => Promise.reject(Error(messages.errors.unhandledResponse)));

/**
 * fetch.gradeOverrideHistory(subsectionId, userId)
 * fetches grade override history for a given user on a given subsection
 * @param {string} subsectionId - subsection identifier
 * @param {string} userId - user identifier
 * @return {Promise} - get response
 */
const gradeOverrideHistory = (subsectionId, userId) => (
  get(sectionOverrideHistoryUrl(subsectionId, userId))
);

/*********************************************************************************
 * POST Actions
 *********************************************************************************/
/**
 * updateGradebookData(updateData)
 * sends an update message with new grades overrides
 * @param {object[]} updateData
 *  {
 *    user_id: <int>,
 *    usage_id: <string>
 *    grade: {
 *      earned_all_override: <int>
 *      possible_all_override: <int>
 *      earned_graded_override: <int>
 *      possible_graded_override: <int>
 *    }
 *  }
 * @return {Promise} - post response
 */
const updateGradebookData = (updateData) => post(urls.getBulkUpdateUrl(), updateData);

/**
 * uploadGradeCsv(formData)
 * Posts form data to grade csv url.  On success, forwards response data.
 * Reject promise with result on failure.
 * @param {object} formData - new grade data
 * @return {Promise} - post response
 */
const uploadGradeCsv = (formData) => (
  post(gradeCsvUrl(), formData).then((result) => {
    if (result.status === 200 && !result.data.error_messages.length) {
      return result.data;
    }
    return Promise.reject(result);
  })
);

export default StrictDict({
  fetch: StrictDict({
    assignmentTypes,
    cohorts,
    gradebookData,
    gradeBulkOperationHistory,
    gradeOverrideHistory,
    roles,
    tracks,
  }),
  updateGradebookData,
  uploadGradeCsv,
});
