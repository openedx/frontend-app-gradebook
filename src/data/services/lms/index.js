import { StrictDict } from '@src/utils';
import api from './api';
import urls from './urls';

export default StrictDict({
  api,
  urls,
});
