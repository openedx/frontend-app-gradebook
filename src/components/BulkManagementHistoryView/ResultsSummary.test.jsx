import React from 'react';
import { shallow } from 'enzyme';

import { Icon } from '@edx/paragon';
import { Download } from '@edx/paragon/icons';

import lms from 'data/services/lms';
import ResultsSummary from './ResultsSummary';

jest.mock('@edx/paragon', () => ({
  Hyperlink: () => 'Hyperlink',
  Icon: () => 'Icon',
}));
jest.mock('@edx/paragon/icons', () => ({
  Download: 'DownloadIcon',
}));
jest.mock('data/services/lms', () => ({
  urls: {
    bulkGradesUrlByRow: jest.fn((rowId) => ({ url: { rowId } })),
  },
}));

describe('ResultsSummary component', () => {
  const props = {
    rowId: 42,
    text: 'texty',
  };
  let el;
  const assertions = [
    'safe hyperlink with bulkGradesUrl with course and row id',
    'download icon',
    'results text',
  ];
  beforeEach(() => {
    el = shallow(<ResultsSummary {...props} />);
  });
  test(`snapshot - ${assertions.join(', ')}`, () => {
    expect(el).toMatchSnapshot();
  });
  test('Hyperlink has target="_blank" and rel="noopener noreferrer"', () => {
    expect(el.props().target).toEqual('_blank');
    expect(el.props().rel).toEqual('noopener noreferrer');
  });
  test('Hyperlink has href to bulkGradesUrl', () => {
    expect(el.props().href).toEqual(lms.urls.bulkGradesUrlByRow(props.rowId));
  });
  test('displays Download Icon and text', () => {
    const icon = el.childAt(0);
    expect(icon.is(Icon)).toEqual(true);
    expect(icon.props().src).toEqual(Download);
    expect(el.childAt(1).text()).toEqual(props.text);
  });
});
