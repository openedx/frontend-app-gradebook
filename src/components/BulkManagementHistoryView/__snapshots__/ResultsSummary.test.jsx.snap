// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsSummary component snapshot - safe hyperlink with bulkGradesUrl with course and row id, download icon, results text 1`] = `
<Hyperlink
  destination="www.edx.org"
  href={
    Object {
      "url": Object {
        "rowId": 42,
      },
    }
  }
  rel="noopener noreferrer"
  showLaunchIcon={false}
  target="_blank"
>
  <Icon
    className="d-inline-block"
    src="DownloadIcon"
  />
  texty
</Hyperlink>
`;
